lua: 5.2 compatibility and other fixes

grondo opened the pull request:
This PR updates the code in `src/modules/wreck/luastack.c` for compatibility with Lua 5.2.

Additionally, a minor testing fix is included (discovered on c9.io), along with the removal of accidentally committed debug code that used `inspect.lua`, which we do not install with the flux-core package.

Finally, Travis testing is updated to run at least one builder using lua5.2, to ensure continued support of that version. (have yet to see if that works)

garlick commented:
Restarted the first two builders again.  They had stalled in the same place as before (right after `rolemask/loop.t`).

garlick commented:
Strange, those two are stuck in the same place again.  I'll hold off rerunning in case travis is sad right now.

grondo commented:
Ok, retrying without the travis-ci changes

grondo commented:
without the lua 5.2 changes, a different test hung in a different place. Trying again with the travis-ci changes. If the first 2 builders hang again I'm giving up for now.

grondo commented:
Hangs again in the same spot. Let me try one more thing (not even sure my travis changes were working)

grondo commented:
Still hanging. Let me make sure this is not a problem with the use of lua 5.2 before trying anything else

grondo commented:
Ok, I think I found the issue. The version of luarocks in Ubuntu 14.04 only works with Lua 5.1, so the built luaposix module is also for 5.1, which the lua5.2 interpreter will happily load, but then hangs --  which happens in the `sharness.d/flux-sharness.sh` code here:
```sh
#  Test requirements for testsuite
if ! lua -e 'require "posix"'; then
    error "failed to find lua posix module in path"
fi
```

I'll try to find a different way to test 5.2.

grondo commented:
Ok, after many, many, many attempts, and a few fixes along the way, I think I have this PR testing against Lua 5.1 and 5.2. The changes are:

 - Don't install the lua apt packages, instead use [hererocks](https://github.com/mpeterv/hererocks), which is a helper script that installs Lua to separate roots, including a copy of luarocks. `hererocks` is installed by pip and used in the `travis-dep-builder.sh`
 - Use Lua 5.1 by default, otherwise check `$LUA_VERSION` exported from the travis-ci build matrix.
 - I had a problem where it seemed like the build matrix env vars (except CC) were not exported to the `before_install` script, so I moved everything in `travis.yml` to just the `script`
 - Other than that, there were just a few places where `#!/usr/bin/lua` was still hard coded, and some baffling `LUA_PATH` and `LUA_CPATH` issues that I think I've worked out for travis anyway.

grondo commented:
It occurs to me that the `travis-dep-builder.sh` changes here will affect flux-sched builds, which use the same script.

I should probably make the `hererocks` setup optional so that it is not active by default.

grondo commented:
Updated the `travis-dep-builder.sh` script so that it only uses `hererocks` when `LUA_VERSION` is set or the `-L, --lua-version` option is used. This should make it so no changes are required in `flux-sched` once this PR is merged.

I also figured out why I thought environment variables from the .travis.yml weren't being set in the `before_install` script (my fault), so I removed the changes to move everything into `script:`, which reduced the amount of churn in the PR.

We'll see if this actually builds though...

grondo commented:
Sadly, now getting python test errors in the LUA_VERSION=5.2 builds only. :angry:

grondo commented:
I was able to get some more information by setting `SIDEFLUX_DEBUG` in the travis environment and the failure above is what you get when `flux start` fails. In this case (not surprisingly) this was a Lua issue. I believe the problem is some Lua modules are being linked statically with liblua.a, and Lua 5.2 especially doesn't like this when liblua.so is also dynamically linked.

```
#Running doctests and unit tests in: /home/travis/build/grondo/flux-core/flux-core-0.7.0-1887-g797b3840/src/bindings/python/test/handle.py
flux-start: warning: setting --bootstrap=selfpmi due to --size option

2018-07-21T22:53:02.612216Z broker.err[0]: rc1: lua: multiple Lua VMs detected

2018-07-21T22:53:02.612345Z broker.err[0]: rc1: stack traceback:

2018-07-21T22:53:02.612364Z broker.err[0]: rc1: 	[C]: ?

2018-07-21T22:53:02.612368Z broker.err[0]: rc1: 	[C]: in function 'require'

2018-07-21T22:53:02.612371Z broker.err[0]: rc1: 	...887-g797b3840/_build/sub/../../src/cmd/flux-hostlist:30: in main chunk

2018-07-21T22:53:02.612376Z broker.err[0]: rc1: 	[C]: ?

2018-07-21T22:53:02.617863Z broker.err[0]: rc1: lua: multiple Lua VMs detected

2018-07-21T22:53:02.617980Z broker.err[0]: rc1: stack traceback:

2018-07-21T22:53:02.618012Z broker.err[0]: rc1: 	[C]: ?

2018-07-21T22:53:02.618037Z broker.err[0]: rc1: 	[C]: in function 'require'

2018-07-21T22:53:02.618065Z broker.err[0]: rc1: 	...887-g797b3840/_build/sub/../../src/cmd/flux-hostlist:30: in main chunk

2018-07-21T22:53:02.618102Z broker.err[0]: rc1: 	[C]: ?

2018-07-21T22:53:02.618363Z broker.err[0]: Run level 1 Exited with non-zero status (rc=1) 0.4s

flux-start: 0 (pid 21411) exited with rc=1

    return self.gen.next()
  File "/home/travis/build/grondo/flux-core/flux-core-0.7.0-1887-g797b3840/src/bindings/python/test_commands/sideflux.py", line 130, in run_beside_flux
    f.start()
  File "/home/travis/build/grondo/flux-core/flux-core-0.7.0-1887-g797b3840/src/bindings/python/test_commands/sideflux.py", line 92, in start
    raise EnvironmentError(self.sub.poll())
EnvironmentError: 1
close failed in file object destructor:
IOError: [Errno 9] Bad file descriptor
Exception OSError: (3, 'No such process') in <bound method SideFlux.__del__ of <sideflux.SideFlux object at 0x2ad0215aead0>> ignored
```


I'm not sure why this was a problem in the sideflux environment only (all other tests pass), but the problem was due to some lua APT packages still being installed in Travis. Once I removed those, the issue went away.

grondo commented:
Thanks!
